OpenACC Model

pranav-sivaraman opened the pull request:
We have created an OpenACC implementation of Cloverleaf based on the OpenMP Offload Implementation

cc: @jhdavis8

## Implementation Notes

### Pragma Conversions

| OpenMP | OpenACC |
| ------------- | ------------- |
| `omp target`  | `acc`  |
| `teams distribute parallel for simd`  | `parallel loop gang worker vector`  |
| `map(alloc : ...)` | `create(...)`
| `from(...)` | `host(...)` (this gets converted to self)
| `use_device_ptr(...)` | `deviceptr(...)`
| `map(release:...)` | `delete(...)`

For every accelerated loop there is also a `default(present)` clause to indicate that the data is on the GPU and not generate any extra copies.

### Function Conversions
| OpenMP | OpenACC |
| ------------- | ------------- |
| `omp_get_num_devices()`  | `acc_get_num_devices(acc_device_default)`  |
| `omp_set_default_device()` | `acc_set_device_num(..., acc_device_default)` |
| `omp_get_default_device()` | `acc_get_device_num(acc_device_default)` |

Tested with NVHPC@22.7 on V100 and A100 and got similar performance results to the OpenMP version.

Any feedback would be greatly appreciated

pranav-sivaraman committed:
feat: initial OpenACC model setup

pranav-sivaraman committed:
feat: add OpenACC model to CMake

pranav-sivaraman committed:
feat: convert OpenMP pragmas to OpenACC

pranav-sivaraman committed:
style: license

pranav-sivaraman committed:
fix: C++17 standard for OpenACC

pranav-sivaraman committed:
feat: clean up OpenACC data

pranav-sivaraman committed:
feat: add MPI support for OpenACC model

pranav-sivaraman committed:
fix: OpenACC data clauses

tom91136 commented:
Thanks for the PR! I'll review it shortly. Is there a RAJA version as well? I saw a RAJA commit from your fork which looks very interesting.

pranav-sivaraman commented:
Yes we have a RAJA version as well!

Right now I am having issues just getting stuff to consistently build with the whole RAJA suite (RAJA/CHAI,Umpire) across all of our RAJA implementations. Especially if the RAJA suite is externally installed (with Spack). So we are waiting on getting feedback from the RAJA developers before we proceed with any PRs involving RAJA.

tom91136 commented on src/acc/calc_dt.cpp:
```diff
@@ -0,0 +1,156 @@
+  double *xvel0 = field.xvel0.data;
+  double *yvel0 = field.yvel0.data;
+
+/* TODO I think this loop has room for speedup? Instead of doing 1 do 5 min reductions? */
```
You mean `reduction(min: dtct, dtut, dtvt, dtdivt, dt_min_va)` and then a final `fmin(dt_min_val, fmin(dtct, fmin(dtut, fmin(dtvt, fmin(dtdivt, g_big)))));` on the host? I would be interested to see the difference if you have time to benchmark this, preferably using `clover_bm16` or larger, ideally  `clover_bm64` or if memory allows `clover_bm256`.

pranav-sivaraman replied:
Yes that's what I would like to try. Thanks for the suggestions for the input sizes. I will definitely contact you with any findings I have.

tom91136 commented on src/acc/generate_chunk.cpp:
```diff
@@ -0,0 +1,167 @@
+
+/* TODO NVHPC says it can't parallelize the inner for loops
+ * We can try forcing if the index calcs are unique */
+#pragma acc parallel loop gang worker vector default(present) collapse(2) clover_use_target(globals.context.use_target)                            \
```
In this particular case I'm not too worried because it's only used a few times for initialisation so we're just ensuring the data resides at the right place.
Please elaborate on what you mean by forcing it.

tom91136 commented on src/acc/pack_kernel.cpp:
```diff
@@ -0,0 +1,329 @@
+  double *left_snd = left_snd_buffer.data;
+  double *field = field_buffer.data;
+  const size_t field_sizex = field_buffer.nX();
+/* TODO NVHPC Compiler is also saying that it can't parallelize this loop
```
This one is slightly more concerning, does `collapse(2)` help? 
One of the OpenMP target implementations we've tested (it may have been vanilla Clang) had some performance issues with `collapse(2)`  in packing/unpacking so we didn't write it that way.

pranav-sivaraman replied:
I am not exactly sure what the impact of `collapse(2)` will have on performance. I would probably need to do profiling to fully investigate. It does pass validation with `clover_bm16_short.in` if I do add a `collapse(2)`. 

We could add the `collapse(2)` just for OpenACC since NVHPC seems to be the only compiler with a mature implementation. 

As for the OpenMP version maybe there are performance issues with vanilla Clang due to `simd` not being fully implemented in `teams distribute parallel for simd`? I think we should add the `collapse(2)` to the OpenMP version if I am assuming the NVHPC compiler does not have performance issues. Hopefully vanilla clang's OpenMP implementation progresses where this won't become an issue in the future?

tom91136 replied:
Yes, please profile and keep `collapse(2)` if it helps. Open a separate PR if you think OpenMP will also benefit from this and the compiler doesn't fail or given wrong results.

tom91136 commented on src/acc/comms_kernel.cpp:
```diff
@@ -0,0 +1,266 @@
+    create(right_rcv[ : right_rcv_buffer.N()]) create(right_snd[ : right_snd_buffer.N()])                                        \
+    create(top_rcv[ : top_rcv_buffer.N()]) create(top_snd[ : top_snd_buffer.N()])                                                \
+    create(bottom_rcv[ : bottom_rcv_buffer.N()]) create(bottom_snd[ : bottom_snd_buffer.N()])
+  {
```
We have a bug here if we 1) use a CUDA-aware MPI or 2) specify `--staging true`  in the CLI and have >1 MPI ranks.
On NVHPC 23.9 (CUDA 12.2) with the bundled HPC-X MPI, there's a ~35% deviation from the solution on `clover_bm16_short.in`.

I'm not too well versed in OpenACC, but to pass validation, I made the following changes:

1. Replace this directive with `#pragma acc enter data`, drop the extra `{}` (`acc exit data` stays the same).
2. Replace all uses of `#pragma acc data if (!stage) deviceptr(...)` with `#pragma acc host_data if (!stage) use_device(...)`, as we're doing an CUDA-aware MPI call in the block, not a device-side kernel.

Can you double check whether you can reproduce this?

pranav-sivaraman replied:
Are you sure you get a ~35% deviation when you specify `--staging-buffer true`? I get that ~35 deviation when I do a `--staging-buffer false`. I have attached my [output](https://github.com/UoB-HPC/CloverLeaf/files/13233350/output.txt).

Otherwise your changes did indeed pass validation for both `true` and `false` for `--staging-buffer` on `clover_bm16_short.in`.

tom91136 replied:
Sorry, it was a typo, `--staging-buffer true` forces device-to-host staging so it's the safe path.

tom91136 reviewed (changes requested):
Thanks again for the PR, this is looking good.
Just a few minor TODOs and the bug in `comms_kernel`. 
Once resolved, please retest with >1 MPI ranks (can be on the same GPU, performance will be bad)  and try both `--staging true` and `--staging false`.

pranav-sivaraman committed:
fix: staging for MPI calls

pranav-sivaraman commented on src/acc/generate_chunk.cpp:
```diff
@@ -0,0 +1,167 @@
+            if (vertexy[j + 1] >= state_ymin[state] && vertexy[j] < state_ymax[state]) {
+              energy0[i + j * base_stride] = state_energy[state];
+              density0[i + j * base_stride] = state_density[state];
+              for (int kt = j; kt <= j + 1; ++kt) {
```
```suggestion
              #pragma acc loop collapse(2)
              for (int kt = j; kt <= j + 1; ++kt) {
```
Since NVHPC could not parallelize the following loops from its own analysis we can "force" the compiler to parallelize the loops by adding a  `#pragma acc loop collapse(2)` to all the nested for loops. 

I passed validation on `clover_bm16_short.in`, but I am not sure if other input decks may result in an error.

tom91136 replied:
Be aware that only the following decks have checksums:
```
clover_bm.in
clover_bm16.in
clover_bm16_short.in
clover_bm16_very_short.in
clover_bm2_short.in
clover_bm4_short.in
clover_bm_short.in 
```
In general, if `clover_bm16.in` (this one has 3000 steps so is good at catching memory leaks and MPI related issues) is good at 1 and > 1 MPI ranks with and without staging, then the implementation is correct.

pranav-sivaraman commented:
I've implemented the bug fixes regarding `comms_kernel` and have verified that it's passing the chosen test.

I would like add an option for using managed memory for OpenACC, but NVHPC also supports it for OpenMP as well. Should this be a separate PR? If it is would it make sense to remove the CPU options for OpenACC? TBH I am not too well versed on the CPU options so maybe instead just refactor to make it like OMP Offload? That way we can also add flags for `gcc` as well for both AMD and NVIDIA gpus.

tom91136 commented:
Yes, please open a separate PR for changes outside of OpenACC. I think the current OpenACC implementation works fine on the host, but feel free to make any changes you think are appropriate and I'll review it.

pranav-sivaraman committed:
style: remove TODO for PR merge

pranav-sivaraman committed:
feat: add managed memory option for OpenACC

pranav-sivaraman commented:
@tom91136 I've added a managed memory option for OpenACC. As of right now I think its fine to merge in. Any other improvements would probably come with seperate PR.

One Note:
I've tested this with `gcc` OpenACC offloading and there appears to be a bug with the `default(present)` . I think the fix is replace all the `default(present)` with explicit present clauses. Not sure if it's worth doing since the better performance ends up lying in `NVHPC` anyways.

pranav-sivaraman committed:
feat: add offload flags option to allow user to supply flags

pranav-sivaraman committed:
refactor: add present clauses for all kernels

GCC 13 does not implement the "default(present)" clause. Need to add explicit
present clauses to work.

pranav-sivaraman commented:
I've made some change to get the OpenACC model to work with GCC. I think this is ready to merge, but let me know if there are any other changes that need to be made.

pranav-sivaraman committed:
revert: reduction clause changes

NVHPC fails clover_bm16.in with the previous changes. However, GCC now gets a
runtime error when executing calc_dt.cpp

pranav-sivaraman commented:
Hmm after doing some testing on the `clover_bm16.in` input deck I noticed some changes causing NVHPC to fail. I have since reverted those and it is now passing, but GCC seems to be getting a runtime error in `calc_dt.cpp`

I get the following warning when compiling with GCC using the following flags `-fopenacc -foffload=nvptx-none='-march=sm_80 -foffload=-lm -foffload=-O3`
```
[1/2] Building CXX object CMakeFiles/cloverleaf.dir/src/acc/calc_dt.cpp.o
../src/acc/calc_dt.cpp: In function 'calc_dt_kernel(bool, int, int, int, int, double, double, double, double, double, field_type&, double&, int&, double&, double&, int&, int&, int&) [clone ._omp_fn.0]':
../src/acc/calc_dt.cpp:60:9: warning: 'dt_min_val.32' is used uninitialized [-Wuninitialized]
   60 | #pragma acc parallel loop gang worker vector collapse(2) clover_use_target(use_target) \
      |         ^~~
../src/acc/calc_dt.cpp:60:9: note: 'dt_min_val.32' was declared here
   60 | #pragma acc parallel loop gang worker vector collapse(2) clover_use_target(use_target) \
      |         ^~~
../src/acc/calc_dt.cpp:60:9: warning: 'dt_min_val.6' is used uninitialized [-Wuninitialized]
   60 | #pragma acc parallel loop gang worker vector collapse(2) clover_use_target(use_target) \
      |         ^~~
../src/acc/calc_dt.cpp:60:9: note: 'dt_min_val.6' was declared here
   60 | #pragma acc parallel loop gang worker vector collapse(2) clover_use_target(use_target) \
``` 

The runtime error is `CUDA Exception: Warp Illegal Address`. 

I'm not exactly sure what is the exact issue. Could it be related to some alignment issues? 

Edit: I've now fixed the issue by doing something similar to what was done in OpenMP Offload

pranav-sivaraman committed:
fix: reduction clause for calc_dt

tom91136 commented:
@pranav-sivaraman Sorry for the delay! I think this is ready for merge. Just to check, without  `present` , GCC still fails right? From what I understand, this works on GCC, and NVHPC currently.

pranav-sivaraman commented:
With `default(present)`, `gcc@13` and before will fail. Adding the explicit `present` will cause it not to throw a runtime error. `gcc@14` says in their release notes they have now implemented `default(present)` for OpenMP/OpenACC, but I have not tested this.

tom91136 commented:
Thank you very much for this contribution @pranav-sivaraman and for putting up with this long delay. 
This brings us one step closer to the coverage BabelStream has.
